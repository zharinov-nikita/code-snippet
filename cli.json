[
  {
    "template": "ui",
    "files": ".cli/ui"
  },
  {
    "template": "page",
    "files": ".cli/page"
  }
]
