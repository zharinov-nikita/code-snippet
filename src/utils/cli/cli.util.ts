import chalk from 'chalk'
import fs from 'fs'
import recursiveReaddir from 'recursive-readdir'
import { UtilConfig, UtilPath, UtilPrompt, UtilSnippet } from '..'
import type { TypeArgs, TypeConfig } from '../../types'

export class UtilCli {
  private readonly snippet: UtilSnippet
  private readonly config: TypeConfig
  private readonly prompt: UtilPrompt
  private readonly path: UtilPath
  private args: TypeArgs

  constructor() {
    this.snippet = new UtilSnippet()
    this.prompt = new UtilPrompt()
    this.config = new UtilConfig().generate()
    this.path = new UtilPath()
    this.args = {
      snippet: '',
      name: '',
      path: '',
      isFlat: false,
    }
  }

  public async start(): Promise<void> {
    try {
      const args = await this.prompt.getArgs()
      this.args = { ...this.args, ...args }
      const { snippet, name, path, isFlat } = this.args

      const config = this.config.find((item) => item.snippet === snippet)
      if (!config) throw new Error('config')

      if (snippet === '') throw new Error(`specify the required argument snippet=example`)
      if (name === '') throw new Error(`specify the required argument name=example`)
      if (path === '') throw new Error(`specify the required argument path=example`)

      recursiveReaddir(config.files, (error, paths) => {
        if (error) throw new Error(error.message)

        paths.forEach((pathToSnippet) => {
          const formattedSnippet = this.snippet.format({ pathToSnippet, name })
          const generatePath = this.path.generate({
            pathToSnippet,
            config,
            name,
            path,
          })
          if (fs.existsSync(generatePath.pathWriteSnippetFile)) throw new Error(name)
          this.snippet.create({ ...generatePath, formattedSnippet, isFlat })
        })
      })
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(chalk.red(e))
    }
  }
}
