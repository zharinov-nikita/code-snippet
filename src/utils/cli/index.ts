export * from './cli.type'
export * from './cli.util'
