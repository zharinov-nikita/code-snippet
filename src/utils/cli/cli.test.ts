import { describe, expect, test } from '@jest/globals'
import { UtilCli } from './cli.util'

describe('util/cli', () => {
  const util = new UtilCli()

  test('snapshot', () => {
    expect(util).toMatchSnapshot('util')
  })
})
