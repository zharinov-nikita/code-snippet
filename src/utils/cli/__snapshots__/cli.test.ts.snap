// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/cli snapshot: util 1`] = `
UtilCli {
  "args": {
    "isFlat": false,
    "name": "",
    "path": "",
    "snippet": "",
  },
  "config": [
    {
      "files": ".create-snippet\\my-snippet",
      "snippet": "my-snippet",
    },
  ],
  "path": UtilPath {
    "string": UtilString {},
  },
  "prompt": UtilPrompt {
    "config": UtilConfig {},
  },
  "snippet": UtilSnippet {
    "string": UtilString {},
  },
}
`;
